src/coreclr/tools/Common/System/Collections/Generic/ArrayBuilder.cs

```diff
@@ -29,6 +29,19 @@ public void Add(T item)
             _items[_count++] = item;
         }
 
+#if NET
+        public readonly Span<T> AsSpan(int start) => _items.AsSpan(start);
+
+        public Span<T> GetSpan(int size)
```

Comment:
Should we follow the established IBufferWriter<T> pattern - `GetSpan(int sizeHint = 0)` method and `Advance` method to commit the write?

Reply:
That would probably be the better pattern for `GetSpan` overload. My current usage of it is more akin to `ValueStringBuilder.AppendSpan`, i.e.
https://github.com/dotnet/runtime/blob/69431607285e03cd6dd648c55b9ce0f9b95620d8/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L266-L277

I ran bench for the BufferWriter pattern. I left one validation branch in the [Advance](https://github.com/PaulusParssinen/ArrayBuilderBench/blob/f4560024baeb2976b5626a082e0b60de53cb2f35/ArrayBuilder.cs#L36-L50), following `ArrayBufferWriter<T>`'s implementation.

| Method     | Version      | N    |        Mean | Code Size |
| ---------- | ------------ | ---- | ----------: | --------: |
| EmitUShort | "GetSpan"    | 10   |    45.45 ns |     616 B |
| EmitUShort | BufferWriter | 10   |    41.85 ns |     655 B |
|            |              |      |             |           |
| EmitUInt   | "GetSpan"    | 10   |    47.95 ns |     606 B |
| EmitUInt   | BufferWriter | 10   |    42.64 ns |     660 B |
|            |              |      |             |           |
| EmitLong   | "GetSpan"    | 10   |    50.88 ns |     636 B |
| EmitLong   | BufferWriter | 10   |    47.50 ns |     669 B |
|            |              |      |             |           |
| EmitUShort | "GetSpan"    | 100  |   155.59 ns |     632 B |
| EmitUShort | BufferWriter | 100  |   164.55 ns |     679 B |
|            |              |      |             |           |
| EmitUInt   | "GetSpan"    | 100  |   179.09 ns |     633 B |
| EmitUInt   | BufferWriter | 100  |   183.69 ns |     675 B |
|            |              |      |             |           |
| EmitLong   | "GetSpan"    | 100  |   202.89 ns |     650 B |
| EmitLong   | BufferWriter | 100  |   222.30 ns |     657 B |
|            |              |      |             |           |
| EmitUShort | "GetSpan"    | 1000 | 1,111.93 ns |     614 B |
| EmitUShort | BufferWriter | 1000 | 1,429.73 ns |     665 B |
|            |              |      |             |           |
| EmitUInt   | "GetSpan"    | 1000 | 1,219.34 ns |     626 B |
| EmitUInt   | BufferWriter | 1000 | 1,550.34 ns |     672 B |
|            |              |      |             |           |
| EmitLong   | "GetSpan"    | 1000 | 1,442.24 ns |     632 B |
| EmitLong   | BufferWriter | 1000 | 1,817.28 ns |     673 B |

Thoughts on whether we should rename the current method to VSB's `AppenSpan` or add `Advance` to follow IBufferWriter pattern? 🤔

Reply:
I do not have strong opinion about AppendSpan vs. GetSpan+Advance.

I have commented on this because GetSpan that includes advance looked a bit non-intuitive to me.
 
> . I left one validation branch in the [Advance](https://github.com/PaulusParssinen/ArrayBuilderBench/blob/f4560024baeb2976b5626a082e0b60de53cb2f35/ArrayBuilder.cs#L36-L50),

For internal helper types like this one, we typically do debug-only validation with asserts.

Reply:
It is a bit non-intuitive and a pattern that wouldn't get through a public API review 😄 

However as this type is not looking to replace the existing `ArrayBufferWriter<T>`, I think I personally prefer the `Span<byte> AppendSpan(int size)` which is often what one wants given they already have to calculate the minimum size their for their data. I haven't yet personally had use for the extra space that the `Span<byte> IBufferWriter.GetSpan(int hint)` API may give you. Also with it, you need to _remember_ to call. 

With that rationale I'll propose renaming the current method to follow ValueStringBuilder's `AppendSpan` pattern.

Reply:
Sounds good to me.